src/safeposix/net.rs

```diff
@@ -27,51 +51,124 @@ pub fn mux_port(addr: interface::GenIpaddr, port: u16, domain: i32, istcp: bool)
-                self.used_port_set.insert(porttuple);
-                return Ok(port);
+
+        //start from the starting location we specified in a previous attempt to get an ephemeral port
+        let next_ephemeral = if domain == AF_INET {self.next_ephemeral_port_udpv4} else if domain == AF_INET6 {self.next_ephemeral_port_udpv6} else {unreachable!()};
+        for range in [(EPHEMERAL_PORT_RANGE_START ..= next_ephemeral), (next_ephemeral + 1 ..= EPHEMERAL_PORT_RANGE_END)] {
+            for ne_port in range.rev() {
+                let port = ne_port.to_be(); //ports are stored in network endian order
+                porttuple.1 = port;
+
+                //if we think we can bind to this port
+                if !self.port_in_use(&porttuple) {
+                    self.used_port_set.insert(porttuple, if rebindability {1} else {0}); //rebindability of 0 means not rebindable, 1 means it's rebindable and there's 1 bound to it
+
+                    if ne_port == EPHEMERAL_PORT_RANGE_START {
+                        if domain == AF_INET {
+                            self.next_ephemeral_port_udpv4 = EPHEMERAL_PORT_RANGE_END;
+                        } else if domain == AF_INET6 {
+                            self.next_ephemeral_port_udpv6 = EPHEMERAL_PORT_RANGE_END;
+                        } else {unreachable!()};
+                    } else {
+                        if domain == AF_INET {
+                            self.next_ephemeral_port_udpv4 = ne_port - 1;
+                        } else if domain == AF_INET6 {
+                            self.next_ephemeral_port_udpv6 = ne_port - 1;
+                        } else {unreachable!()};
+                    }
+
+                    return Ok(port);
+                }
             }
         }
         return Err(syscall_error(Errno::EADDRINUSE, "bind", "No available ephemeral port could be found"));
     }
-    pub fn _get_available_tcp_port(&mut self, addr: interface::GenIpaddr, domain: i32) -> Result<u16, i32> {
+    pub fn _get_available_tcp_port(&mut self, addr: interface::GenIpaddr, domain: i32, rebindability: bool) -> Result<u16, i32> {
+        if !NET_DEVICES_LIST.contains(&addr) {
+            return Err(syscall_error(Errno::EADDRNOTAVAIL, "bind", "Specified network device is not set up for lind or does not exist!"));
+        }
         let mut porttuple = mux_port(addr.clone(), 0, domain, TCPPORT);
-        for port in (EPHEMERAL_PORT_RANGE_START ..= EPHEMERAL_PORT_RANGE_END).rev().map(|x| x.to_be()) {
-            porttuple.1 = port;
-            if !self.port_in_use(&porttuple) {
-                self.used_port_set.insert(porttuple);
-                return Ok(port);
+
+        //start from the starting location we specified in a previous attempt to get an ephemeral port
+        let next_ephemeral = if domain == AF_INET {self.next_ephemeral_port_tcpv4} else if domain == AF_INET6 {self.next_ephemeral_port_tcpv6} else {unreachable!()};
```

Comment:
This code block and the above block for udp seem extremely similar? Does it make sense to condense this into a function?

Reply:
They do but it's not that easy as if you look inside the loop they have to use different members (next_ephemeral_port_tcpv4/v6 or next_ephemeral_port_udpv4/v6)